language/diem-vm/parallel-executor/src/executor.rs

```diff
@@ -0,0 +1,185 @@
+    sync::Arc,
+};
+
+pub struct ParallelTransactionExecutor<T: Transaction, E: ExecutorTask, I: ReadWriteSetInferencer> {
+    num_cpus: usize,
+    inferencer: I,
+    phantom: PhantomData<(T, E, I)>,
+}
+
+impl<T, E, I> ParallelTransactionExecutor<T, E, I>
+where
+    T: Transaction,
+    E: ExecutorTask<T = T>,
+    I: ReadWriteSetInferencer<T = T>,
+{
+    pub fn new(inferencer: I) -> Self {
+        Self {
+            num_cpus: num_cpus::get(),
+            inferencer,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn execute_transactions_parallel(
+        &self,
+        task_initial_arguments: E::Argument,
+        signature_verified_block: Vec<T>,
+    ) -> Result<Vec<E::Output>, E::Error> {
+        let num_txns = signature_verified_block.len();
+        let chunks_size = max(1, num_txns / self.num_cpus);
+
+        // Get the read and write dependency for each transaction.
+        let infer_result: Vec<_> = {
+            match signature_verified_block
+                .par_iter()
+                .with_min_len(chunks_size)
+                .map(|txn| {
+                    Ok((
+                        self.inferencer.infer_reads(txn)?,
+                        self.inferencer.infer_writes(txn)?,
+                    ))
+                })
+                .collect::<AResult<Vec<_>>>()
+            {
+                Ok(res) => res,
+                // Inferencer passed in by user failed to get the read/writeset of a transaction,
+                // abort parallel execution.
+                Err(_) => return Err(Error::InferencerError),
+            }
+        };
+
+        // Use write analysis result to construct placeholders.
+        let path_version_tuples: Vec<(T::Key, usize)> = infer_result
+            .par_iter()
+            .enumerate()
+            .with_min_len(chunks_size)
+            .fold(Vec::new, |mut acc, (idx, (_, txn_writes))| {
+                acc.extend(txn_writes.clone().into_iter().map(|ap| (ap, idx)));
+                acc
+            })
+            .flatten()
+            .collect();
+
+        let (versioned_data_cache, max_dependency_level) =
+            MVHashMap::new_from_parallel(path_version_tuples);
+        let outcomes = OutcomeArray::new(num_txns);
+
+        let scheduler = Arc::new(Scheduler::new(num_txns));
+
+        scope(|s| {
+            // How many threads to use?
+            let compute_cpus = min(1 + (num_txns / 50), self.num_cpus - 1); // Ensure we have at least 50 tx per thread.
+            let compute_cpus = min(num_txns / max_dependency_level, compute_cpus); // Ensure we do not higher rate of conflict than concurrency.
+
+            for _ in 0..(compute_cpus) {
+                s.spawn(|_| {
+                    let scheduler = Arc::clone(&scheduler);
+                    // Make a new executor per thread.
+                    let task = E::init(task_initial_arguments);
+
+                    while let Some(idx) = scheduler.next_txn_to_execute() {
```

Comment:
Are you gonna add logic to not break out when work isn't immediately available in this diff or later? I think without it we have observed variance & perf drop in some experiments (maybe more in the other context) - so would require convincing benchmarks with many read dependencies to tell it's safe to return and not loop around until everything is done. (that's also the only case when spin loop hint is needed i think).